server/services/matchingQueue.js

```diff
@@ -0,0 +1,96 @@
+const Queue = require("bull");
+const { matchingQueries } = require("../database");
+const { assignThread, unassignThread } = require("../controllers/user");
+const { addToNoAssign, updateStatus } = require("../controllers/thread");
+const config = require("../config/config");
+const matchingQueue = new Queue("candidate matching");
+
+/* job object:
+    {
+        thread: instance of Thread model
+        currentAssignee: user_id of assignee previously set from the result obj
+        pass: number of matching passes (not implemented)
+    }
+
+    result object:
+    {
+        assignee: user_id of assigned user, null if assignment failed + error message will be populated
+    }
+*/
+
+matchingQueue.process(async (job, done) => {
+  const thread = job.data.thread;
+  try {
+    if (thread.status === config.server.threadStatus.indexOf("ongoing")) {
+      throw new Error("Thread already has a reviewer");
+    } else if (job.data.currentAssignee) {
+      // if true, thread had been assigned previously and is now declined or timed-out
+      // remove thread from the current assignee
+      unassignThread(job.data.currentAssignee, thread._id);
+    }
+    const candidates = await matchingQueries.getCandidates(thread);
+    if (candidates.length === 0) {
+      throw new Error("No candidates found");
```

Comment:
are we doing the more general queries in this PR or is that something later (can make sense for something later)

Reply:
Right now it's just matching to the first person who has space (less than 5 assigned) and has higher experience in that language. I was thinking we would do additional queries later.